openapi/src/main/java/io/helidon/openapi/ExpandedTypeDescription.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2019-2020 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package io.helidon.openapi;
+
+import java.beans.IntrospectionException;
+import java.beans.PropertyDescriptor;
+import java.lang.annotation.Annotation;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.microprofile.openapi.models.Extensible;
+import org.eclipse.microprofile.openapi.models.media.Schema;
+import org.yaml.snakeyaml.TypeDescription;
+import org.yaml.snakeyaml.error.YAMLException;
+import org.yaml.snakeyaml.introspector.MethodProperty;
+import org.yaml.snakeyaml.introspector.Property;
+import org.yaml.snakeyaml.introspector.PropertySubstitute;
+import org.yaml.snakeyaml.introspector.PropertyUtils;
+import org.yaml.snakeyaml.nodes.Node;
+import org.yaml.snakeyaml.nodes.ScalarNode;
+
+/**
+ * Extension of {@link TypeDescription} that handles:
+ * <ul>
+ *     <li>nested enums,</li>
+ *     <li>extensible types, and</li>
+ *     <li>references.</li>
+ * </ul>
+ * <p>
+ *     The OpenAPI document format uses lower-case enum names and values, while the SmallRye
+ *     definitions use upper-case. This class simplifies adding the special handling for enums
+ *     declared within a particular class.
+ * </p>
+ * <p>
+ *     Some of the MP OpenAPI items are extensible, meaning they accept sub-item keys with the
+ *     "x-" prefix. This class supports extensions. For scalars it delegates to the normal
+ *     SnakeYAML processing to correctly type and parse the scalar. For sequences it
+ *     creates {@code List}s. For mappings it creates {@code Map}s. The subnodes of the lists and
+ *     maps are handled by the normal SnakeYAML parsing, so the resulting elements in lists and
+ *     maps are of the SnakeYAML-inferred types.
+ * </p>
+ * <p>
+ *     A subnode {@code $ref} maps to the {@code ref} property on the MP OpenAPI types. This type
+ *     description simplifies defining the {@code $ref} property to those types that support it.
+ * </p>
+ * <p>
+ *     We use this expanded version of {@code TypeDescription} with the generated SnakeYAMLParserHelper class.
+ * </p>
+ */
+class ExpandedTypeDescription extends TypeDescription {
```

Comment:
Can this be `final`?

Reply:
This one cannot be `final`; the inner `SchemaTypeDescription` class extends it. But _that_ one can be `final`; I changed that.